### PR TITLE
Simplify serving files and cache

### PR DESCRIPTION
**Pull request by bdraco**
Followup to address https://github.com/hacs/integration/pull/1655#issuecomment-748382518

### CODE REVIEW

**Comment by quthla** [comment]
I verified that setting `X-Accel-Expires: @1` makes nginx cache the items and send 304.

Can you add that header to the PR?

**Comment by bdraco** [comment]
> I verified that setting `X-Accel-Expires: @1` makes nginx cache the items and send 304.
> 
> 
> 
> Can you add that header to the PR? 

Would you please open a new PR for this since it's nginx specific and not directly related to what this PR is solving.

File: custom_components/hacs/helpers/functions/file_etag.py [hunk 1]
```diff
@@ -2,20 +2,19 @@
...
 
 async def async_get_etag(path) -> bool:
     hass = get_hacs().hass
-    fnv = await hass.async_add_executor_job(get_etag, path)
-    if fnv is None:
+    etag = await hass.async_add_executor_job(get_etag, path)
+    if etag is None:
         return None
-    return str(hex(fnv))
+    return '"' + str(hex(etag)) + '"'
```
Comment (ludeeus) [inline custom_components/hacs/helpers/functions/file_etag.py:20]: Why is this needed? `str()`  makes it a string right?

Comment (bdraco) [reply custom_components/hacs/helpers/functions/file_etag.py:20]: Firefox actually wants the whole string in quotes

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag

**Comment by quthla** [comment]
I'm afraid there's still an issue with your (original) ETag PR.

```
< HTTP/1.1 304 Not Modified
< Content-Length: 0
< Content-Type: application/octet-stream
```

The latter two apparently must equal to those of the original response. It's not compliant with the RFCs otherwise (and causes issues with caching)

https://github.com/apache/httpd/blob/trunk/modules/cache/mod_cache.c#L1185

`cache: https://hass:443/hacsfiles/mini-media-player/mini-media-player-bundle.js? responded with an uncacheable 304, retrying the request /hacsfiles/mini-media-player/mini-media-player-bundle.js. Reason: contradiction: 304 Not Modified; but Content-Length, Content-Type modified`

**Comment by bdraco** [comment]
> I'm afraid there's still an issue with your (original) ETag PR.
> 
> 
> 
> ```
> 
> < HTTP/1.1 304 Not Modified
> 
> < Content-Length: 0
> 
> < Content-Type: application/octet-stream
> 
> ```
> 
> 
> 
> The latter two apparently must equal to those of the original response. It's not compliant with the RFCs otherwise (and causes issues with caching)
> 
> 
> 
> https://github.com/apache/httpd/blob/trunk/modules/cache/mod_cache.c#L1185
> 
> 
> 
> `cache: https://hass:443/hacsfiles/mini-media-player/mini-media-player-bundle.js? responded with an uncacheable 304, retrying the request /hacsfiles/mini-media-player/mini-media-player-bundle.js. Reason: contradiction: 304 Not Modified; but Content-Length, Content-Type modified`

I think you are referring to RFC7230 3.3.2:

A server MAY send a Content-Length header field in a 304 (Not Modified) response to a conditional GET request (Section 4.1 of [RFC7232]); a server MUST NOT send Content-Length in such a response unless its field-value equals the decimal number of octets that would have been sent in the payload body of a 200 (OK) response to the same request.


I think we can drop the content type and content length headers when we send a 304.  That should happen in another PR as it's not related to solving the performance issue in this PR.

**Comment by bdraco** [comment]
I looks like this issue here https://github.com/aio-libs/aiohttp/issues/4777

**Comment by bdraco** [comment]
> It looks like this issue here https://github.com/aio-libs/aiohttp/issues/4777

Looking though the code, there doesn't appear to be a way to turn off the content type default.  We will probably have to call the mime type detector code and set it manually

**Comment by bdraco** [comment]
It looks like we can actually simplify this quite a bit since `no-cache` and `max-age=0, must-revalidate` are functionally equivalent according to: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#Requiring_revalidation

Since we are not preserving `mtime` when installing files, it looks like `if-modified-since` will actually be fine to use which means we can use upstream support and remove all the etag code.

**Comment by bdraco** [comment]
@quthla Does the latest push solve all issues from your perspective except for the `X-Accel-Expires: @1` header for `nginx` ?

**Comment by bdraco** [comment]
Verified upgrade/downgrade behavior working as expected.

**Comment by bdraco** [comment]
Verified we still get a 304 on non modified files.

**Comment by quthla** [comment]
Apache still not caching it but can't find reason why in log anymore. Anyways, I had already disabled caching before for /hacsfiles so I would at least get 304 instead of 200 full body from Apache.

For me it's good to go.

**Comment by bdraco** [comment]
Reverified that its working as expected when simple thermostat updated today

Screen Shot 2020-12-29 at 2 08 15 PM
Screen Shot 2020-12-29 at 2 08 12 PM

**Review by ludeeus — APPROVED** [review]
